src/main/java/de/dhbw/karlsruhe/grammar/generation/GrammarConcatenationGeneration.java

```diff
@@ -0,0 +1,111 @@
+    private List<GrammarProduction> grammarRules;
+    private String[] terminals;
+    private String[] nonTerminals;
+    private Random random = new Random();
+    private GrammarConcatenationVerification grammarConcatenationVerification;
+
+    public Grammar generateGrammar() {
+        grammarConcatenationVerification = new GrammarConcatenationVerification();
+
+        do {
+            this.startGeneration();
+        } while (!grammarConcatenationVerification.verifyProductions(this.grammarRules, this.nonTerminals));
+
+        return new Grammar(this.terminals, this.nonTerminals, this.grammarRules.toArray(GrammarProduction[]::new), this.nonTerminals[0]);
+    }
+
+    private void startGeneration () {
+        this.grammarRulesSet = new HashSet<>();
+        int anzTerminals = this.random.nextInt(4) + 1;
+        int anzNonTerminals = this.random.nextInt(7) +1;
+
+        this.addTerminals(anzTerminals);
+        this.addNonTerminals(anzNonTerminals);
+        this.grammarRules = this.generateProductions();
+    }
+
+    private void addNonTerminals(int pAnzNonTerminals) {
+        Set<String> nonTerminalsSet = new HashSet<>();
+        char nonTerminal;
+
+        do {
+            nonTerminal = RandomStringUtils.random(1, true, false).toUpperCase().charAt(0);
+            nonTerminalsSet.add(Character.toString(nonTerminal));
+        } while(nonTerminalsSet.size() < pAnzNonTerminals);
+
+        this.nonTerminals = nonTerminalsSet.toArray(String[]::new);
+    }
+
+    private void addTerminals(int pAnzTerminals) {
+        Set<String> terminalsSet = new HashSet<>();
+        char terminal;
+
+        do {
+            terminal = RandomStringUtils.random(1, true, true).toLowerCase().charAt(0);
+            terminalsSet.add(Character.toString(terminal));
+        } while(terminalsSet.size() < pAnzTerminals);
+
+        this.terminals = terminalsSet.toArray(String[]::new);
+    }
+
+    protected List<GrammarProduction> generateProductions() {
+        String rightSide = "";
+
+        for (String nonTerminal : nonTerminals) {
+            int anzProductions = this.random.nextInt(4) +1 ;
+            int setSizeBeforeCurrentGeneration = this.grammarRulesSet.size();
+            do {
+                do {
+                    rightSide = generateRightSide();
+                } while (rightSide.equalsIgnoreCase(nonTerminal));
+
+                GrammarProduction gr = new GrammarProduction(nonTerminal, rightSide);
+                this.grammarRulesSet.add(gr);
+            } while ((this.grammarRulesSet.size() - setSizeBeforeCurrentGeneration) < anzProductions);
+        }
+
+        Set<String> notTerminatingNonTerminals = this.grammarConcatenationVerification.getNonTerminatingNonTerminals(this.grammarRulesSet, this.nonTerminals);
+
+        List<String> temp = new ArrayList<>(Arrays.stream(this.terminals).toList());
+        temp.add("epsilon");
+        String[] terminalsAndEpsilon = temp.toArray(new String[0]);
```

Comment:
Erst Liste erstellen, dann epsilon hinzufügen und dann wieder zu Array konvertieren ist nicht nötig. Direkt ein neues Array erstellen mit Länge der Terminale+1 usw.

Reply:
Ist das nicht aufwändiger, weil ich jedes Element erstmal in das neue Array kopieren muss?

Reply:
Ist das nicht aufwändiger, weil ich jedes Element erstmal in das neue Array kopieren muss?

Reply:
Arraycopy musst du nicht manuell machen. Sowas wäre mein Vorschlag:
![image](https://user-images.githubusercontent.com/92317814/225014660-b8a709f8-06cb-4ab7-816b-962c3ad7f755.png)